[libcxx] improves diagnostics for containers with bad value types

cjdb opened the pull request:
`std::vector<int&>` generates nearly 170 lines of diagnostic, most of which are redundant, and it's only helpful if you are already familiar with the rule that containers can't hold reference types. This commit reduces it to only a handful of lines, all of which are dedicated to clearly communicating the problem. It also applies this to all the other containers, and for all non-cv-qualified and non-object types.

These static_asserts are placed at the very top of each container because they short-circuit further instantiation errors, thereby leading a smaller set of diagnostics for the programmer. Placing them in `std::allocator` (which is the common denominator for all containers) doesn't do the short circuiting, and we thus end up with several unhelpful diagnostics after the helpful one.

This commit only cleans up things that are already diagnosed as compile-time errors. In particular, `std::map<int&&, int>` should be ill-formed, per [associative.reqmts.general]/p8. libc++ and libstdc++ currently permit that, but not `std::map<int&, int>`. As such, this commit only adds a static assert for `map<K&, V>`, and defers diagnosing valid productions that should be ill-formed to a future commit.

cjdb committed:
[libcxx] improves diagnostics for containers with bad value types

`std::vector<int&>` generates nearly 170 lines of diagnostic, most of
which are redundant, and it's only helpful if you are already familiar
with the rule that containers can't hold reference types. This commit
reduces it to only a handful of lines, all of which are dedicated to
clearly communicating the problem. It also applies this to all the
other containers, and for all non-cv-qualified and non-object types.

These static_asserts are placed at the very top of each container
because they short-circuit further instantiation errors, thereby
leading a smaller set of diagnostics for the programmer. Placing them
in `std::allocator` (which is the common denominator for all containers)
doesn't do the short circuiting, and we thus end up with several
unhelpful diagnostics after the helpful one.

This commit only cleans up things that are already diagnosed as
compile-time errors. In particular, `map<int&&, int>` should be
ill-formed, per [associative.reqmts.general]/p8. libc++ and libstdc++
currently permit that, but not `map<int&, int>`. As such, this
commit only adds a static assert for `map<K&, V>`, and defers diagnosing
valid productions that should be ill-formed to a future commit.

cjdb commented:
The motivation for this patch comes from routinely getting downstream user bugs about the helpfulness of the current diagnostics.

ldionne commented on libcxx/include/forward_list:
```diff
@@ -635,8 +640,136 @@ void __forward_list_base<_Tp, _Alloc>::clear() _NOEXCEPT {
+
 template <class _Tp, class _Alloc /*= allocator<_Tp>*/>
 class _LIBCPP_TEMPLATE_VIS forward_list : private __forward_list_base<_Tp, _Alloc> {
+  static_assert(!is_const<_Tp>::value, "'std::forward_list' can only hold non-const types");
```
While I think the `static_assert`s definitely make sense (in some form, maybe grouped under a common utility), I really don't think the added complexity is worth it for the `__forward_list_base` specializations above.

What is the reason for needing those? If the reason is that the compiler otherwise won't skip emitting spurious diagnostics, IMO I would treat it as a compiler problem. In particular, we could inherit from `__forward_list_value_type_constraints<_Tp>` and have the `static_assert`s in there, and somehow have Clang stop trying to compile junk when it encounters a hard error like a failed `static_assert` (I never really understood why it tries to keep going).

cjdb replied:
Would it be possible to temporarily ship with this complexity, until we get Clang into a good state? Solving this in libc++ doesn't let the compiler off the hook, because it's going to be prevalent in other template libraries, but we can at least offer our users a bit of relief in the meantime. Here are the diagnostics issued when (a) we use what's proposed in this PR, (b) we delete my `__forward_list_base` specialisations, and (c) we put the static_asserts in `__forward_list_base`.

**Code under test**
```cpp
#include <forward_list>

std::forward_list<int&> fl;
```

**Checks go in `forward_list` + `__forward_list_base` specialisations**
```
/home/cjdb/projects/llvm-project/build/include/c++/v1/forward_list:770:17: error: static assertion failed due to requirement '!is_reference<int &>::value': 'std::forward_list' can only hold object types; references are not objects
  770 |   static_assert(!is_reference<_Tp>::value, "'std::forward_list' can only hold object types; references are not objects");
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
hello.cpp:3:25: note: in instantiation of template class 'std::forward_list<int &>' requested here
    3 | std::forward_list<int&> x;
      |                         ^
1 error generated.
```

**Checks go in `forward_list` only**
```
In file included from hello.cpp:1:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:209:
/home/cjdb/projects/llvm2/build/include/c++/v1/__memory/allocator.h:87:17: error: static assertion failed due to requirement '!is_reference<int &>::value': 'std::allocator' can only allocate object types; references are not objects
   87 |   static_assert(!is_reference<_Tp>::value,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/cjdb/projects/llvm2/build/include/c++/v1/__memory/allocator_traits.h:248:39: note: in instantiation of template class 'std::allocator<int &>' requested here
  248 |   using value_type         = typename allocator_type::value_type;
      |                                       ^
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:488:20: note: in instantiation of template class 'std::allocator_traits<std::allocator<int &>>' requested here
  488 |   typedef typename allocator_traits<allocator_type>::void_pointer void_pointer;
      |                    ^
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:644:51: note: in instantiation of template class 'std::__forward_list_base<int &, std::allocator<int &>>' requested here
  644 | class _LIBCPP_TEMPLATE_VIS forward_list : private __forward_list_base<_Tp, _Alloc> {
      |                                                   ^
hello.cpp:3:25: note: in instantiation of template class 'std::forward_list<int &>' requested here
    3 | std::forward_list<int&> x;
      |                         ^
In file included from hello.cpp:1:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:199:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/__algorithm/lexicographical_compare.h:14:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/__algorithm/mismatch.h:16:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/__algorithm/unwrap_iter.h:14:
/home/cjdb/projects/llvm2/build/include/c++/v1/__memory/pointer_traits.h:177:1: error: no member named 'rebind' in 'std::pointer_traits<int>'
  177 | using __rebind_pointer_t = typename pointer_traits<_From>::template rebind<_To>;
      | ^~~~~
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:315:1: note: in instantiation of template type alias '__rebind_pointer_t' requested here
  315 | using __begin_node_of = __forward_begin_node<__rebind_pointer_t<_VoidPtr, __forward_list_node<_Tp, _VoidPtr> > >;
      | ^
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:490:11: note: in instantiation of template type alias '__begin_node_of' requested here
  490 |   typedef __begin_node_of<value_type, void_pointer> __begin_node;
      |           ^
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:644:51: note: in instantiation of template class 'std::__forward_list_base<int &, std::allocator<int &>>' requested here
  644 | class _LIBCPP_TEMPLATE_VIS forward_list : private __forward_list_base<_Tp, _Alloc> {
      |                                                   ^
hello.cpp:3:25: note: in instantiation of template class 'std::forward_list<int &>' requested here
    3 | std::forward_list<int&> x;
      |                         ^
In file included from hello.cpp:1:
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:648:17: error: static assertion failed due to requirement '!is_reference<int &>::value': 'std::forward_list' can only hold object types; references are not objects2
  648 |   static_assert(!is_reference<_Tp>::value, "'std::forward_list' can only hold object types; references are not objects2");
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
hello.cpp:3:25: note: in instantiation of template class 'std::forward_list<int &>' requested here
    3 | std::forward_list<int&> x;
      |                         ^
3 errors generated.
```

**Checks go in `__forward_list_base`**
```
In file included from hello.cpp:1:
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:487:17: error: static assertion failed due to requirement '!is_reference<int &>::value': 'std::forward_list' can only hold object types; references are not objects2
  487 |   static_assert(!is_reference<_Tp>::value, "'std::forward_list' can only hold object types; references are not objects2");
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:650:51: note: in instantiation of template class 'std::__forward_list_base<int &, std::allocator<int &>>' requested here
  650 | class _LIBCPP_TEMPLATE_VIS forward_list : private __forward_list_base<_Tp, _Alloc> {
      |                                                   ^
hello.cpp:3:25: note: in instantiation of template class 'std::forward_list<int &>' requested here
    3 | std::forward_list<int&> x;
      |                         ^
In file included from hello.cpp:1:
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:652:26: error: no type named '__node_allocator' in 'std::__forward_list_base<int &, std::allocator<int &>>'
  652 |   typedef typename base::__node_allocator __node_allocator;
      |           ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
hello.cpp:3:25: note: in instantiation of template class 'std::forward_list<int &>' requested here
    3 | std::forward_list<int&> x;
      |                         ^
In file included from hello.cpp:1:
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:653:26: error: no type named '__node_type' in 'std::__forward_list_base<int &, std::allocator<int &>>'
  653 |   typedef typename base::__node_type __node_type;
      |           ~~~~~~~~~~~~~~~^~~~~~~~~~~
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:654:26: error: no type named '__node_traits' in 'std::__forward_list_base<int &, std::allocator<int &>>'
  654 |   typedef typename base::__node_traits __node_traits;
      |           ~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:655:26: error: no type named '__node_pointer' in 'std::__forward_list_base<int &, std::allocator<int &>>'
  655 |   typedef typename base::__node_pointer __node_pointer;
      |           ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:656:26: error: no type named '__begin_node_pointer' in 'std::__forward_list_base<int &, std::allocator<int &>>'
  656 |   typedef typename base::__begin_node_pointer __begin_node_pointer;
      |           ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
In file included from hello.cpp:1:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:209:
/home/cjdb/projects/llvm2/build/include/c++/v1/__memory/allocator.h:87:17: error: static assertion failed due to requirement '!is_reference<int &>::value': 'std::allocator' can only allocate object types; references are not objects
   87 |   static_assert(!is_reference<_Tp>::value,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/cjdb/projects/llvm2/build/include/c++/v1/__memory/allocator_traits.h:248:39: note: in instantiation of template class 'std::allocator<int &>' requested here
  248 |   using value_type         = typename allocator_type::value_type;
      |                                       ^
/home/cjdb/projects/llvm2/build/include/c++/v1/__memory/allocator_traits.h:378:66: note: in instantiation of template class 'std::allocator_traits<std::allocator<int &>>' requested here
  378 |   static_assert(is_same<_Alloc, __rebind_alloc<_Traits, typename _Traits::value_type> >::value,
      |                                                                  ^
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:662:17: note: in instantiation of template class 'std::__check_valid_allocator<std::allocator<int &>>' requested here
  662 |   static_assert(__check_valid_allocator<allocator_type>::value, "");
      |                 ^
hello.cpp:3:25: note: in instantiation of template class 'std::forward_list<int &>' requested here
    3 | std::forward_list<int&> x;
      |                         ^
In file included from hello.cpp:1:
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:664:62: error: no type named 'value_type' in 'std::allocator<int &>'
  664 |   static_assert(is_same<value_type, typename allocator_type::value_type>::value,
      |                                     ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
hello.cpp:3:25: note: in instantiation of template class 'std::forward_list<int &>' requested here
    3 | std::forward_list<int&> x;
      |                         ^
In file included from hello.cpp:1:
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:677:26: error: no type named 'iterator' in 'std::__forward_list_base<int &, std::allocator<int &>>'
  677 |   typedef typename base::iterator iterator;
      |           ~~~~~~~~~~~~~~~^~~~~~~~
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:678:26: error: no type named 'const_iterator' in 'std::__forward_list_base<int &, std::allocator<int &>>'
  678 |   typedef typename base::const_iterator const_iterator;
      |           ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
In file included from hello.cpp:1:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:199:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/__algorithm/lexicographical_compare.h:13:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/__algorithm/min.h:14:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/__algorithm/min_element.h:16:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/__functional/invoke.h:14:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/__type_traits/invoke.h:15:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/__type_traits/decay.h:15:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/__type_traits/is_array.h:14:
In file included from /home/cjdb/projects/llvm2/build/include/c++/v1/cstddef:37:
/home/cjdb/projects/llvm2/build/include/c++/v1/__type_traits/enable_if.h:28:58: error: no type named 'type' in 'std::enable_if<false, int>'; 'enable_if' cannot be used to disable this declaration
   28 | using __enable_if_t _LIBCPP_NODEBUG = typename enable_if<_Bp, _Tp>::type;
      |                                                          ^~~
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:694:13: note: in instantiation of template type alias '__enable_if_t' requested here
  694 |   template <__enable_if_t<__is_allocator<_Alloc>::value, int> = 0>
      |             ^
hello.cpp:3:25: note: in instantiation of template class 'std::forward_list<int &>' requested here
    3 | std::forward_list<int&> x;
      |                         ^
In file included from hello.cpp:1:
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:789:30: error: multiple overloads of 'push_front' instantiate to the same signature 'void (int &)'
  789 |   _LIBCPP_HIDE_FROM_ABI void push_front(const value_type& __v);
      |                              ^
hello.cpp:3:25: note: in instantiation of template class 'std::forward_list<int &>' requested here
    3 | std::forward_list<int&> x;
      |                         ^
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:787:30: note: previous declaration is here
  787 |   _LIBCPP_HIDE_FROM_ABI void push_front(value_type&& __v);
      |                              ^
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:809:34: error: multiple overloads of 'insert_after' instantiate to the same signature 'iterator (const_iterator, int &)' (aka 'int (int, int &)')
  809 |   _LIBCPP_HIDE_FROM_ABI iterator insert_after(const_iterator __p, const value_type& __v);
      |                                  ^
/home/cjdb/projects/llvm2/build/include/c++/v1/forward_list:804:34: note: previous declaration is here
  804 |   _LIBCPP_HIDE_FROM_ABI iterator insert_after(const_iterator __p, value_type&& __v);
      |                                  ^
13 errors generated.
```

Each has its drawbacks:

* (a) has the clearest diagnostic, but adds a bit of source complexity for libc++ maintainers
* (b) is not particularly viable, because it talks about `allocator` rather than `forward_list`, and not everyone knows that `Container<T>` is a shorthand for `Container<T, allocator<T>>` (this is why I didn't update just `allocator`).
* (c) is the simplest for maintainers, but produces the most text.

@AaronBallman, do you know how difficult it would be to get Clang to achieve (a)'s diagnostic when the static_assert is embedded in a base class?

AaronBallman replied:
> (I never really understood why it tries to keep going).

It's a tradeoff. Follow-on diagnostics can sometimes be frustratingly chatty and useless but getting one error at a time is also frustrating. So we try to recover from errors by getting back into a valid state (e.g., unknown type? let's pretend it was `int` instead.) in an effort to get more relevant diagnostics. But, it has its problems.

> @AaronBallman, do you know how difficult it would be to get Clang to achieve (a)'s diagnostic when the static_assert is embedded in a base class?

I don't have a good intuition for how hard it would be (this involves template instantiation machinery on top of the usual sema machinery). I suspect it will be somewhat challenging to implement without degrading diagnostic behavior in other cases. However, it does seem reasonable to explore if someone has the bandwidth to experiment with changing the diagnostic recovery behavior in that case.

AaronBallman replied:
> I've very often thought to myself that the diagnostics quality would be greatly improved if we just had -ferror-limit=1 be the default.

Personal opinion: I think that's a worse user experience in a lot of cases; much of the follow-on diagnostic problems are language-specific (C++ and templates are not known for excellent diagnostic behavior). Trickling out the errors one at a time may be reasonable when doing many quick edit/recompile cycles with small changes where it's unlikely to add multiple errors, but may not be reasonable for someone throwing a giant pile of code at the compiler one time. I think it would be more palatable to take a bunch of cases where we're emitting too many follow-on diagnostics to see if we can identify a pattern, and reduce the number of follow-ons in that specific case. e.g., we may decide "oh, class template instantiation issue? Let's just bail out early instead of marking the node as invalid and continuing".

Lead maintainer opinion: that would have to go through an RFC to see whether the community would accept it or not.

cjdb replied:
> The best diagnostics seem to be the ones we have when we put the checks in `forward_list` and use `-ferror-limit=1`, basically.

We need to put the checks in `__forward_list_base` since putting them in `forward_list` without the helpers means the diagnostic mentions `std::allocator` instead of `std::forward_list` when `-ferror-limit=1` (the `forward_list`-specific diagnostic is the third error). tl;dr: examples (1) and (3) are good; (2) is there just to show what happens.

> @cjdb As I said before, I think the static assertions are fine, however I don't think trying to work around `-ferror-limit` is reasonable. It's unfortunately an infinitely deep can of worms since basically every place where we have a `static_assert` has the exact same problem. Furthermore, as Clang changes (or with different compilers), our attempts to silence spurious diagnostics could end up not working or simply producing different diagnostics -- basically I think that's a very brittle construct to maintain.

I'd like to discuss this more generally (preferably on Discourse) before settling on this as a blanket policy. I don't think it's worth blocking this PR on that discussion's resolution, so I've applied your changes.

Related, but any resolution is beyond the scope of this patch: is `__forward_list_base` actually necessary? I've only skimmed it, but can't see anything that would require `forward_list` to need a base class (similarly for `std::list`'s base).

ldionne replied:
> Related, but any resolution is beyond the scope of this patch: is __forward_list_base actually necessary? I've only skimmed it, but can't see anything that would require forward_list to need a base class (similarly for std::list's base).

It might not be necessary, but one reason I've seen these base classes used for is exception safety. `forward_list` constructs an instance of the base class and then does something else, and if something goes wrong in there, the base class will be properly destroyed. IDK if that's the case here but that's the first thing I'd look out for if I was trying to refactor this.

cjdb replied:
I think I've completed the actionable bits for this PR.

philnik777 commented on libcxx/include/__memory/allocator.h:
```diff
@@ -76,8 +79,16 @@ struct __non_trivial_if<true, _Unique> {
 class _LIBCPP_TEMPLATE_VIS allocator : private __non_trivial_if<!is_void<_Tp>::value, allocator<_Tp> > {
-  static_assert(!is_const<_Tp>::value, "std::allocator does not support const types");
-  static_assert(!is_volatile<_Tp>::value, "std::allocator does not support volatile types");
+  static_assert(!is_const<_Tp>::value, "'std::allocator' can only allocate non-const object types");
+  static_assert(!is_volatile<_Tp>::value, "'std::allocator' can only allocate non-volatile object types");
+  static_assert(!is_reference<_Tp>::value || !is_function<typename remove_reference<_Tp>::type>::value,
+                "'std::allocator' can only allocate object types; function references are not objects (consider using "
+                "a function pointer)");
+  static_assert(!is_reference<_Tp>::value,
+                "'std::allocator' can only allocate object types; references are not objects");
```
Could we just check `is_same<remove_cvref_t<_Tp>, _Tp>` instead?

cjdb replied:
That erases the tailored diagnostics; noting `references are not objects`, etc., are hints I that believe significantly improves UX.

cjdb replied:
Folks internally suggested that I change the text to the following, as they feel it'll be more helpful than what I'm currently proposing.

* `std::allocator<T const>`: `'std::allocator' cannot allocate const objects` (return to status quo)
* `std::allocator<T volatile>`: `'std::allocator' cannot allocate volatile objects` (return to status quo)
* `std::allocator<T&>`: `'std::allocator' cannot allocate references`
* `std::allocator<T()>`: `'std::allocator' cannot allocate functions`

We do lose the function reference suggestion, but I'm okay with doing that for something that's more helpful.

philnik777 replied:
Why not simply "cannot allocate cv or reference qualified objects"? I'm not sure how much of an improvement it is to specify the qualifier explicitly, since it's already shown in the diagnostic.

cjdb replied:
I kicked the internal discussion off by asking if the average C++ programmer would be able to understand ` 'std::vector' can only allocate non-cv-qualified object types`, and the feedback I've received can be aggregated as roughly

* most C++ programmers don't know what the term "cv-qualified" means
* we should only mention relevant information, and mentioning "volatile" when `volatile` isn't used was repeatedly called out in particular
* most C++ programmers probably aren't going to know that "object" has a very different meaning to other languages
* most non-toolchain folks in the discussion seemed to prefer separate messages; toolchain folks were sort of ambivalent

What do you feel a single check will offer over independent checks?

philnik777 replied:
When you have a `const int&` you get the information that you have to remove both the `const` and the `&` at once, not only after you've removed one or the other. We also avoid three instantiations for every type we have a `vector` of. Passing this test is the common case after all.

Is cv-qualified really that uncommon/niche? When you look at the cppreference page for `const` you get basically redirected to https://en.cppreference.com/w/cpp/language/cv. FWIW you could also say "const or volatile qualified". It's a bit more verbose, but should be perfectly understandable.

cjdb replied:
I asked in a general C++ chatroom, and ended up with feedback from managers, folks who teach C++ to Googlers, and a handful of ICs. The consensus of the discussion was that "cv-qualified" likely isn't well known, and that specific error messages will be significantly better for the user than a generic one (and especially so if "volatile" is being mentioned in the generic case: that was raised at least twice).

You do raise a good point about `int const&`, and this was also brought up in the discussion. It's certainly frustrating to [get an error, fix it, recompile, and get another error](https://youtu.be/2WZLJpMOxS4?si=9p5lIkYVQ9MhvZ1U). I'd like to address this, and also keep the specific messages. Would you be opposed to additional checks that handle these cases?

philnik777 replied:
I would be against it if that means even more classes being instantiated, but maybe I found something we'd both be happy with. Louis already wants a centralized place for the checks (which I agree would be a good idea), so maybe we could have
```c++
template <class _Tp, bool = is_same<_Tp, __remove_cvref_t<_Tp> >::value>
struct __check_container_constraints {};

template <class _Tp>
struct __check_container_constraints<_Tp, false> {
  static_assert(!is_const<_Tp>::value, "cannot allocate `const` objects");
  static_assert(!is_volatile<_Tp>::value, "Some fancy diagnostic");
  // ...
};
```
That would avoid all the instantiations in the common case. I don't care about the instantiations here if it fails anyways.

cjdb replied:
What is your concern with respect to the number of instantiations?

philnik777 replied:
Well, I'm concerned that we instantiate a bunch of types for the rare case they are actually required for diagnostic purposes. Every instantiation takes time and memory, and for a type that is used everywhere, like `vector`, we should keep unnecessary instantiations as low as possible.

cjdb replied:
It sounds to me like you're concerned about build times. Have you run any preliminary analysis to determine that there is a noticeable impact?

philnik777 replied:
This is more a death by a thousand cuts situation than something significant I think. If we add this many checks everywhere people will definitely notice, but not if we just add it in a single place.

cjdb replied:
Very good point, I agree. I *think* I've got something that bridges the gap between what you suggested above, and what I was originally hoping to land.

cjdb committed:
rewords diagnostics per internal feedback and adds arrays

cjdb committed:
formats files

cjdb committed:
sorts includes missed by clang-format

cjdb committed:
more include sorting

cjdb committed:
more formatting, with a successful git clang-format

I think the best way to enact `git clang-format` might be to use `git
clang-format <hash-before-your-commit-first>` instead of `git
clang-format HEAD~n`. This ensures that all commits in a PR are
formatted, instead of just the last n (which one might miscount, like I
probably did).

cjdb committed:
excludes <array> test from C++03

cjdb committed:
suppresses irrelevant diagnostic

cjdb committed:
s/__is_unbounded_array(T)/__libcpp_is_unbounded_array<T>::value/g

cjdb committed:
applies Louis' request

cjdb committed:
finally gets git-clang-format on side

The trick is to run

```
$ git-clang-format --diff <base-hash> --extensions ,cpp,h | patch -p1
```

ldionne commented on libcxx/include/__memory/allocator.h:
```diff
@@ -76,8 +79,10 @@ struct __non_trivial_if<true, _Unique> {
 class _LIBCPP_TEMPLATE_VIS allocator : private __non_trivial_if<!is_void<_Tp>::value, allocator<_Tp> > {
-  static_assert(!is_const<_Tp>::value, "std::allocator does not support const types");
-  static_assert(!is_volatile<_Tp>::value, "std::allocator does not support volatile types");
+  static_assert(!is_const<_Tp>::value, "'std::allocator' cannot allocate const types");
+  static_assert(!is_volatile<_Tp>::value, "'std::allocator' cannot allocate volatile types");
+  static_assert(!is_reference<_Tp>::value, "'std::allocator' cannot allocate references");
+  static_assert(!is_function<_Tp>::value, "'std::allocator' cannot allocate functions");
```
Can we tie these checks back to the Standard wording that says so? It doesn't have to be in the static assert message, but at least in a comment.

cjdb replied:
PTAL

ldionne commented on libcxx/include/deque:
```diff
@@ -468,6 +475,13 @@ const _DiffType __deque_iterator<_ValueType, _Pointer, _Reference, _MapPointer,
 
 template <class _Tp, class _Allocator /*= allocator<_Tp>*/>
 class _LIBCPP_TEMPLATE_VIS deque {
+  static_assert(!is_const<_Tp>::value, "'std::deque' cannot hold const types");
```
I would like to centralize these checks in some way, since they seem to all basically come from the same container requirements in the standard. If we ever change those requirements, it's almost certain that containers will get out of sync unless we centralize these checks.

cjdb replied:
Working on this. Where is a good spot to put it? I was originally putting it in `__config`, but that's not great IMO. Possibly `__type_traits`, since it only consumes type traits?

philnik777 replied:
I think `__type_traits/diagnostic_utilities.h` or something like that would be good.

ldionne commented on libcxx/test/libcxx/containers/associative/multimap/non_cv_object.verify.cpp:
```diff
@@ -0,0 +1,47 @@
+
+std::multimap<int const, int> K1;
+std::multimap<int, int const> M1;
+// TODO(cjdb): turn this into a compile-time error
```
Instead of adding TODOs here, could you file a Github issue listing the cases you'd like to tackle as a follow-up?

cjdb replied:
How about both? I like doing `// TODO(#<issue-number>)` since it means one can identify exactly where the issue is referring to, and how many problems there are to solve before closing.

cjdb replied:
I've opened #106635.

ldionne reviewed (commented):
I like this patch a lot more in its current form. I still have a few comments but the direction makes sense to me!

cjdb commented:
I think I need to walk back my `is_array` change, since I think libc++ might be incorrectly rejecting `vector<char[2]>`.

cjdb committed:
unifies the containers' diagnostics

A Clang pragma that lets us "hide" macro expansion from diagnostics so
that users don't need to read implementation details would be a good
thing to explore in the future.

ldionne commented on libcxx/include/__type_traits/diagnostic_utilities.h:
```diff
@@ -0,0 +1,47 @@
+// allocator that meets the Cpp17Allocator requirements (https://eel.is/c++draft/allocator.requirements).
+// In particular, this means that containers should only accept non-cv-qualified object types, and
+// types that are Cpp17Erasable.
+#define _LIBCPP_CHECK_CONTAINER_VALUE_TYPE_REQUIREMENTS_BASE(_Container, _Tp)                                          \
```
I don't see this macro being used outside of the definition of `_LIBCPP_CHECK_CONTAINER_VALUE_TYPE_REQUIREMENTS_FULL` -- is that an oversight or should this be inlined into `_LIBCPP_CHECK_CONTAINER_VALUE_TYPE_REQUIREMENTS_FULL`?

cjdb replied:
Ah! It was an attempt to share the content with `std::allocator`, not to handle `std::array` and the map types (the latter of which will be converted after the clean-up).

ldionne commented on libcxx/include/__type_traits/diagnostic_utilities.h:
```diff
@@ -0,0 +1,47 @@
+#include <__type_traits/is_void.h>
+#include <__type_traits/is_volatile.h>
+
+#pragma GCC system_header
```
```suggestion
#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
#  pragma GCC system_header
#endif
```

ldionne commented on libcxx/include/__memory/allocator.h:
```diff
@@ -16,9 +16,12 @@
 #include <__type_traits/is_same.h>
 #include <__type_traits/is_void.h>
 #include <__type_traits/is_volatile.h>
+#include <__type_traits/remove_reference.h>
```
Why is that header added?

ldionne commented on libcxx/include/array:
```diff
@@ -127,11 +127,16 @@ template <size_t I, class T, size_t N> const T&& get(const array<T, N>&&) noexce
+#include <__type_traits/is_unbounded_array.h>
+#include <__type_traits/is_void.h>
 #include <__type_traits/remove_cv.h>
+#include <__type_traits/remove_reference.h>
```
Why is that header added?

ldionne commented on libcxx/include/deque:
```diff
@@ -211,10 +211,18 @@ template <class T, class Allocator, class Predicate>
 #include <__split_buffer>
+#include <__type_traits/diagnostic_utilities.h>
 #include <__type_traits/is_allocator.h>
+#include <__type_traits/is_array.h>
+#include <__type_traits/is_const.h>
 #include <__type_traits/is_convertible.h>
+#include <__type_traits/is_function.h>
+#include <__type_traits/is_reference.h>
 #include <__type_traits/is_same.h>
 #include <__type_traits/is_swappable.h>
+#include <__type_traits/is_void.h>
+#include <__type_traits/is_volatile.h>
+#include <__type_traits/remove_reference.h>
```
I don't think these includes should be necessary. This applies to other headers too.

cjdb committed:
Update libcxx/include/__type_traits/diagnostic_utilities.h

Co-authored-by: Louis Dionne <ldionne.2@gmail.com>

cjdb committed:
removes extraneous headers, adds macro to `std::allocator`

cjdb committed:
applies clang-format

cjdb committed:
changes using `is_array` to `is_bounded_array`

`std::llocator` apparently supports unbounded arrays prior to C++20, so
we need to change the `is_array` check to only check for bounded arrays.

ldionne commented on libcxx/include/__type_traits/diagnostic_utilities.h:
```diff
@@ -0,0 +1,50 @@
+#  define _LIBCPP_CHECK_CONTAINER_VALUE_TYPE_IS_NOT_ARRAY_BEFORE_CXX20(_Template, _Tp, _Verb)
+#else
+#  define _LIBCPP_CHECK_CONTAINER_VALUE_TYPE_IS_NOT_ARRAY_BEFORE_CXX20(_Template, _Tp, _Verb)                          \
+    ;                                                                                                                  \
```
Instead, you can define `_LIBCPP_CHECK_CONTAINER_VALUE_TYPE_IS_NOT_ARRAY_BEFORE_CXX20` to `static_assert(true, "")` and simplify the semi-colon tricks.

ldionne commented on libcxx/include/__type_traits/diagnostic_utilities.h:
```diff
@@ -0,0 +1,50 @@
+
+#define _LIBCPP_CHECK_CONTAINER_VALUE_TYPE_REQUIREMENTS(_Container, _Tp)                                               \
+  static_assert(                                                                                                       \
+      !__libcpp_is_unbounded_array<_Tp>::value, "'std::" _Container "' cannot hold C arrays of an unknown size");      \
```
Instead of passing `vector` as the `_Container` argument, I'd pass `std::vector` and simplify the message here. The message would become `_Container " cannot hold C arrays of an unknown size"`.

cjdb replied:
Hmm, I'd prefer to keep the single quotes so that it's consistent with how Clang outputs type information. For example, this is the diagnostic we get for this erroneous code:

```cpp
namespace std {
  template<class T>
  struct S1;

  template<class T>
  struct S2;
}

std::S1<std::S2> y;
```
```
error: use of class template 'std::S2' requires template arguments
```

This helps us remain more consistent as a whole implementation:

```
error: static assertion failed: 'std::vector' cannot hold C arrays of an unknown size
```

WDYT?

ldionne commented on libcxx/include/module.modulemap:
```diff
@@ -1505,7 +1505,17 @@ module std_private_memory_align                           [system] { header "__m
-module std_private_memory_allocator                       [system] { header "__memory/allocator.h" }
+module std_private_memory_allocator                       [system] {
+  header "__memory/allocator.h"
+  export std_private_type_traits_diagnostic_utilities
+  export std_private_type_traits_is_bounded_array
+  export std_private_type_traits_is_const
+  export std_private_type_traits_is_function
+  export std_private_type_traits_is_reference
+  export std_private_type_traits_is_unbounded_array
+  export std_private_type_traits_is_void
+  export std_private_type_traits_is_volatile
```
Did you run into CI issues if you didn't do that? I don't think these need to be exported from the module since they are only being used in the implementation.

cjdb replied:
Yeah, I'm puzzled too. Our CI build script wasn't happy when building the modules. Might be worth trying on your end to see if I'm doing something silly?

ldionne commented on libcxx/include/unordered_map:
```diff
@@ -1024,6 +1029,15 @@ template <class _Key,
           class _Pred  = equal_to<_Key>,
           class _Alloc = allocator<pair<const _Key, _Tp> > >
 class _LIBCPP_TEMPLATE_VIS unordered_map {
+  // TODO(#106635): replace with _LIBCPP_CHECK_CONTAINER_VALUE_TYPE_REQUIREMENTS
+  // Remember to remove relevant headers when this is completed.
```
I read the description of https://github.com/microsoft/STL/pull/3660 and it looks like the Standard does allow these types to be used. If that's the case, then https://github.com/llvm/llvm-project/issues/106635 should be closed until we have a LWG issue or a paper that changes the Standard, and these TODOs should be removed since we won't be able to address them.

Note that I have no love for reference types in these locations, but I also want to avoid adding non-actionable TODOs.

cjdb replied:
I think MS/STL is allowing these as an extension? The linked wording from #106635 already prohibits `unordered_map<int, int&>`. We weren't catching it before, and have been Hyrum's Law'd.

cjdb replied:
Okay, I get what's happening, thanks to cplusplus/draft#7249. I've parked #106635, and have replaced the TODOs with `// not an error`.

cjdb replied:
Per https://github.com/llvm/llvm-project/issues/106635#issuecomment-2336761764, I've reopened the issue, and have reverted bb82c957a0e57a1d3d1d31ec71020f70ad659dad.

ldionne replied:
So IIUC basically references in these locations are not permitted, and we should clean up the usage.

cjdb replied:
References, `const`, `volatile`, etc. I'm especially leery of `map<int const, int>`, because it's in no way interface-distinguishable from `map<int, int>`, yet is its own instantiation.

I'm happy to take point on that over the next few releases, but we'll need to come up with a plan (and potentially coordinate with MS/STL and libstdc++).

cjdb committed:
reduces the number of trait instantiations

Nikolas pointed out that checking all the traits in all cases will
probably grind builds to a halt as we improve our diagnostics, so this
commit makes a point of reducing the number of instantiations as much as
possible, and preserves the highly specific messages.

cjdb committed:
removes commented out code

cjdb committed:
Merge branch 'main' into cleaner-libcxx-diagnostics

cjdb committed:
post-sync clang-format

cjdb committed:
responds to red CI

cjdb committed:
responds to red CI

cjdb committed:
responds to red CI

cjdb committed:
responds to red CI

cjdb committed:
replaces TODOs

cjdb committed:
Revert "replaces TODOs"

This reverts commit bb82c957a0e57a1d3d1d31ec71020f70ad659dad.

philnik777 commented on libcxx/include/__type_traits/diagnostic_utilities.h:
```diff
@@ -0,0 +1,80 @@
+// allocator that meets the Cpp17Allocator requirements (https://eel.is/c++draft/allocator.requirements).
+// In particular, this means that containers should only accept non-cv-qualified object types, and
+// types that are Cpp17Erasable.
+template <template <class...> class _Template, class _Tp, bool = is_same<typename decay<_Tp>::type, _Tp>::value>
```
```suggestion
template <template <class...> class _Template, class _Tp, bool = is_same<__decay_t<_Tp>, _Tp>::value>
```

philnik777 commented on libcxx/include/__type_traits/diagnostic_utilities.h:
```diff
@@ -0,0 +1,80 @@
+template <template <class...> class _Template, class _Tp, bool = is_same<typename decay<_Tp>::type, _Tp>::value>
+struct __allocator_requirements : true_type {};
+
+#if _LIBCPP_STD_VER >= 20
+template <class _Tp>
+struct __bounded_arrays_allowed_only_after_cxx20 : false_type {};
+#else
+template <class _Tp>
+struct __bounded_arrays_allowed_only_after_cxx20 : integral_constant<bool, __libcpp_is_bounded_array<_Tp>::value> {};
+#endif
```
```suggestion
#if _LIBCPP_STD_VER >= 20
template <class _Tp>
inline const bool __bounded_arrays_allowed_only_after_cxx20 = false;
#else
template <class _Tp>
inline const bool __bounded_arrays_allowed_only_after_cxx20 = __libcpp_is_bounded_array<_Tp>::value;
#endif
```

cjdb replied:
TIL we support inline variables in C++03, but not constexpr ones ☹️ (I do understand why though).

philnik777 commented on libcxx/include/__type_traits/diagnostic_utilities.h:
```diff
@@ -0,0 +1,80 @@
+struct __bounded_arrays_allowed_only_after_cxx20 : integral_constant<bool, __libcpp_is_bounded_array<_Tp>::value> {};
+#endif
+
+#define _LIBCPP_CHECK_ALLOCATOR_VALUE_TYPE_REQUIREMENTS(_Template, _Verb)                                              \
```
We don't have to _Uglify macro parameters.

cjdb committed:
Update libcxx/include/__type_traits/diagnostic_utilities.h

Co-authored-by: Nikolas Klauser <nikolasklauser@berlin.de>

cjdb committed:
Update libcxx/include/__type_traits/diagnostic_utilities.h

Co-authored-by: Nikolas Klauser <nikolasklauser@berlin.de>

cjdb committed:
separates the cv-unqualified object requirement for other diganostics

cjdb committed:
fixes asan diagnostic

ldionne commented on libcxx/include/__type_traits/diagnostic_utilities.h:
```diff
@@ -0,0 +1,94 @@
+
+#define _LIBCPP_DEFINE__REQUIRES_CV_UNQUALIFIED_OBJECT_TYPE(_Template, _Verb)                                          \
+  template <class _Tp>                                                                                                 \
+  struct __requires_cv_unqualified_object_type<_Template, _Tp, false>                                                  \
```
Is there a reason why the patch now defines these `__requires_cv_unqualified_object_type` specializations for each container? Was that in response to the comment about too many instantiations?

This adds complexity but I don't see the benefit (yet), I'd rather keep this patch as simple as can be. After all the goal here is to issue diagnostics -- that's a fairly simple problem and I'd like to keep the solution accordingly simple.

cjdb replied:
> Is there a reason why the patch now defines these `__requires_cv_unqualified_object_type` specializations for each container? Was that in response to the comment about too many instantiations?

Yes. I ended up benchmarking Chrome's build times and while we could probably tolerate this patch in isolation, stacking similar changes will eventually tank build times. 

> This adds complexity but I don't see the benefit (yet), I'd rather keep this patch as simple as can be. After all the goal here is to issue diagnostics -- that's a fairly simple problem and I'd like to keep the solution accordingly simple.

Issuing simple diagnostics is simple. The goal of this patch is to issue diagnostics that are informative to the reader, using language that they're likely to understand; achieving that without adversely impacting build times is apparently more challenging to simplify.

Could you help me understand what it is about the complexity that raises concern, please?

ldionne replied:
Basically, I don't understand why the macros don't expand to a bunch of `static_asserts` directly (like it used to do), instead of going through an additional partial specialization of `__requires_cv_unqualified_object_type` & friends. That's a complex mechanism for doing basically `static_assert(condition)` and I don't understand what we're gaining from that complexity. In fact, I would expect this approach to have more impact on build times since we instantiate more stuff *and* we have a partial specialization.

cjdb replied:
Ah, I get where you're coming from now, thank you. This is a metaprogramming hack to preserve performance for valid use.
* In the case where `decay_t<_Tp>` matches `_Tp`: `__requires_cv_unqualified_type` should only need to instantiate two templates: `decay_t<_Tp>` and `__requires_cv_unqualified_type<_Template, _Tp, true>`.
* In case where `decay_t<_Tp>` is different to `_Tp`: `__requires_cv_unqualified_type` then instantiates all the checks to see why that isn't the case. This is indeed slower, but we're now going for a failed build, and tooling tends to burn compile times in favour of helpful diagnostics.

Neither `__allocator_requirements`, nor `__container_requirements` have the same effect, so I'm happy to revert those without further convincing.

Chromium is a good candidate for benchmarking the worst-case scenario, because it has well over 100k utterances of `std::vector`. Would providing benchmarks be helpful? (It'll take ~a day to get those as I discarded the results from last week, and I tend to run quite a few trials to reduce the impact of noise.)

cjdb replied:
Ping

cjdb replied:
Ping :)

ldionne replied:
I'm skeptical about the performance claims since the current solution requires instantiating `is_same`, `__decay_t`, `__requires_cv_unqualified_object_type` and finding the right specialization for it. That doesn't seem to be significantly less expensive than instantiating the 4 traits directly. And if we wanted to be even speedier, we could use the `_v` versions of these traits (which would bump requirements to C++17), or we could use the compiler builtins directly for `__is_reference` and `__is_function`, which we use unconditionally.

And even if there's a very very small compile-time hit, the complexity introduced by the current approach for what is essentially 4 `static_assert`s seems really difficult to justify.

I would have no opposition to this patch if it was simplified to something like:

```c++
#if _LIBCPP_STD_VER >= 20
# define _LIBCPP_CHECK_ALLOCATOR_REQUIREMENTS(template, value_type)                                                   \
    static_assert(!is_const<value_type>::value, #template " cannot have a const value_type");                         \
    static_assert(!is_volatile<value_type>::value, #template " cannot have a volatile value_type");                   \
    static_assert(!is_reference<value_type>::value, #template " cannot have a value_type that is a reference type");  \
    static_assert(!is_function<value_type>::value, #template " cannot have a value_type that is a function type")
#else
# define _LIBCPP_CHECK_ALLOCATOR_REQUIREMENTS(template, value_type)                                                   \
    static_assert(!is_const<value_type>::value, #template " cannot have a const value_type");                         \
    static_assert(!is_volatile<value_type>::value, #template " cannot have a volatile value_type");                   \
    static_assert(!is_reference<value_type>::value, #template " cannot have a value_type that is a reference type");  \
    static_assert(!is_function<value_type>::value, #template " cannot have a value_type that is a function type");    \
    static_assert(!__libcpp_is_bounded_array<value_type>::value, #template " cannot have a value_type that is a bounded array before C++20")
#endif

#define _LIBCPP_CHECK_ALLOCATOR_AWARE_CONTAINER_REQUIREMENTS(template, value_type)                                    \
  _LIBCPP_CHECK_ALLOCATOR_REQUIREMENTS(template, value_type);                                                         \
  static_assert(!is_void<value_type>::value, #template " cannot hold void");                                          \
  static_assert(!__libcpp_is_unbounded_array<value_type>::value, #template " cannot hold C arrays of an unknown size")
```

I think it's difficult to justify adding more complexity than that either for compilation times or for diagnostics quality. After all, this kind of error is something that users will see once when they try to start using a container wrong, and they will very likely see this near the top of their call stack (i.e. this is unlikely to trigger deep down inside a dependent library). 95% of the benefit of this patch is obtained by adding the `static_assert`s in the simplest way possible, but as it stands a large part of the complexity comes from trying to go the last 5% in terms of diagnostics quality and compilation time (which I'm still skeptical of). I don't think that's the right tradeoff, and that's the source of my pushback.

So, TLDR: I think adding these diagnostics is great and I love that we're making the library more robust to misuse and improving the diagnostics quality. But I don't think the patch achieves the right complexity tradeoff in its current state (it started more naive and with a better tradeoff IMO).

cjdb committed:
Merge branch 'main' into cleaner-libcxx-diagnostics

cjdb committed:
Merge branch 'main' into cleaner-libcxx-diagnostics
